### ISSUE TITLE
Training metrics

### ISSUE THREAD

**Issue by jmarshrossney** [wilsonmr/anvil#19]
From Albergo _et al._
>  In practice, autocorrelations should be evaluated as a test metric alongside the training loss to confirm improvement over the course of training the model.

The `training_metrics` branch introduced (in a very rough-and-ready way) MCMC phases at intervals throughout the training, so the acceptance rate and autocorrelation could be tracked throughout the training.

This can be massively improved upon now that that the restructure has introduced checkpoints for training; we can just iterate over the checkpoints, running `sample`, instead of doing it during the training.

Opening this issue so we can discuss the most sensible way to achieve this. Once we've decided, I'd be happy to code it up. It may already be in the pipeline though!

**Comment by jmarshrossney** [comment]
One thing to note is that we might want to compare like-for-like training between models of different size, in which case we'd want to follow the example of Albergo _et al._ and train models up to a certain acceptance rate, rather than epochs.

This is probably a non-issue for bigger lattices since the acceptance rate will change very slowly anyway, but for smaller ones we could just reduce the number of epochs between checkpoints and pick the one closest to the desired acceptance rate.

**Comment by jmarshrossney** [comment]
Or, do we want to save the acceptance rate and autocorrelation time to a log file, along with the loss, at intervals smaller than the checkpoint intervals?

**Comment by jmarshrossney** [comment]
Just wondering about iterating over checkpoints and running the sampler.
At the moment everything is set up to take one model, run the sampler once, and plot the results.

We might want to separately include the ability to run several sampling simulations for the same model and take the mean, standard deviation.

We might also want to be able to automatically run through every `.pt` file (checkpoint) in a directory, measure the acceptance rate/autocorrelation time etc., and plot the results.

To avoid having to mess around with `phi_four.py` and `config.py` so that it can take lists of checkpoint files, maybe this could be a script job where you give it a directory and it just iterates over all the `.pt` files within, using the runcard to run sampling simulations for each one.

**Comment by jmarshrossney** [comment]
p.s. @wilsonmr is this appropriate use of Github Issues? This is more of a feature request & discussion thread..

**Comment by wilsonmr** [comment]
> p.s. @wilsonmr is this appropriate use of Github Issues? This is more of a feature request & discussion thread..

Yes, fine with me haha

**Comment by wilsonmr** [comment]
> Just wondering about iterating over checkpoints and running the sampler.
> At the moment everything is set up to take one model, run the sampler once, and plot the results.
> 
> We might want to separately include the ability to run several sampling simulations for the same model and take the mean, standard deviation.
> 
> We might also want to be able to automatically run through every `.pt` file (checkpoint) in a directory, measure the acceptance rate/autocorrelation time etc., and plot the results.
> 
> To avoid having to mess around with `phi_four.py` and `config.py` so that it can take lists of checkpoint files, maybe this could be a script job where you give it a directory and it just iterates over all the `.pt` files within, using the runcard to run sampling simulations for each one.

This was something I was playing with over the weekend, I mentioned to @c-j-johnston that there already exists a module which has functionality for looping over configurations and it's called `reportengine`. I actually losely based the restructuring of the project on another thing I've worked on which is based on report engine. I think that it offers us some really nice features:

 - in analysis there are clever ways to loop over things without having to rewrite specific code
 - possibility to make reports which contain a bunch of plots and tables
 - the whole "look before you leap" programming which I implemented in a rushed and stupid way in `config.py` is done in a much nicer way

The downsides are:
 - some of the code gets a bit more complicated and whilst I'm used to the reportengine tools, other people might not be and I don't want that to end up wasting more time than it saves
 - reportengine is still in development so some features could change within the year
 - I couldn't get the reports working properly and I cannot see why

so I wasn't sure exactly how to proceed. perhaps I could try pushing a branch with reportengine implemented and people can see if the code looks utterly crazy or makes sense to them

**Comment by wilsonmr** [comment]
> Or, do we want to save the acceptance rate and autocorrelation time to a log file, along with the loss, at intervals smaller than the checkpoint intervals?

Yeah or a seperate log for each thing, just in case we add a new thing to the log, it doesn't break older logs, they just won't have the logs for new features.. I'm easy either way tbh

**Comment by wilsonmr** [comment]
but with acceptance/autocorrelation, probably we want not check it too frequently, because I guess it takes a while. I guess that would be an argument for just doing it after with checkpoints

**Comment by jmarshrossney** [comment]
> We might want to separately include the ability to run several sampling simulations for the same model and take the mean, standard deviation.

This may be stupid on my part (still in the mindset of local updates and ergodicity issues -> needing lots of separate simulations). We could just run one longer simulation and the statistics will be fine, I think.

**Comment by jmarshrossney** [comment]
Ah I can see why it's a difficult decision whether or not to use reportengine. I'd say if you're interested then go for it.

In the meantime, Cameron and myself can concentrate on the fairly pressing task of reproducing those paper results. I'm sure neither of us will mind manipulating a few files manually if it gets the job done.

**Comment by wilsonmr** [comment]
yeah sounds good

 I can't remember where you told me btw but does @c-j-johnston also know that if you match the action on the arxiv rather than the action in the published version then the observables look a lot nicer?

**Comment by c-j-johnston** [comment]
On the subject of reproducing results, I haven't got numbers which look the same, but I might be using the wrong action -- I've been using whichever is in the `restructure` branch so I'll check if this is the correct one. I've also struggled to get acceptance above 0.7 -- the one model which worked took over 2.5 hours to train, and longer training has not helped.

**Comment by wilsonmr** [comment]
approx 70% is not bad!

**Comment by jmarshrossney** [comment]
This is implemented (though not in the most satisfactory way) by a bash script on the train_to_acceptance branch. Integrating this into normal usage is definitely in the pipeline, and probably doesn't require an outdated Github issue to serve as a reminder.
